### PR TITLE
gh-272 ChainedIterator.getIterator NullPointer Exception.

### LINKED ISSUE #272: ChainedIterable getIterator errors with an Iterable containing nulls.

### PR DESCRIPTION
**Pull request by GCHQDev404**
# Related Issue

- Resolve #272

### CODE REVIEW

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java [hunk 1]
```diff
@@ -63,12 +65,16 @@ private Iterator<? extends T> getIterator() {
         while (!currentIterator.hasNext()) {
             CloseableUtil.close(currentIterator);
             if (iterablesIterator.hasNext()) {
-                currentIterator = iterablesIterator.next().iterator();
+                Object next = iterablesIterator.next();
+                if (nonNull(next) && next instanceof Iterable) {
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:69]: null check not needed as null is not `instanceof Iterable`

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java [hunk 2]
```diff
@@ -63,12 +65,16 @@ private Iterator<? extends T> getIterator() {
...
+                    currentIterator = (Iterator<? extends T>) ((Iterable<?>) next).iterator();
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:70]: Could this also be `(Iterable<? extends T>)` or am I misunderstanding?

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java [hunk 3]
```diff
@@ -63,12 +65,16 @@ private Iterator<? extends T> getIterator() {
...
+                } else if (nonNull(next) && !(next instanceof Iterable)) {
+                    throw new IllegalStateException(String.format("Iterator of Iterator contains non-iterable class: %s object: %s", next.getClass(), next));
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:72]: Should it be `Iterator of Iterables`?

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java [hunk 4]
```diff
@@ -63,12 +65,16 @@ private Iterator<? extends T> getIterator() {
...
+                }
```
Comment (t92549) [inline core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:73]: So if `next` is null we just return `currentIterator`?

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java (hunk 1 above)
Comment (GCHQDev404) [reply core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:69]:
```suggestion
                if (next instanceof Iterable) {
```

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java (hunk 3 above)
Comment (GCHQDev404) [reply core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:72]: The Function required is `KorypheFunction<Iterable<Iterable<I_ITEM>>, Iterable<I_ITEM>>` 
so neither ? :D

Comment (GCHQDev404) [reply core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:72]:
```suggestion
                    throw new IllegalStateException(String.format("Iterable of Iterable contains non-iterable class: %s object: %s", next.getClass(), next));
```

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java (hunk 4 above)
Comment (GCHQDev404) [reply core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:73]: Its hard to see in the view with all the comments but, It should loop back around again to find the next, while there is a next.

File: core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java (hunk 2 above)
Comment (GCHQDev404) [reply core/src/main/java/uk/gov/gchq/koryphe/iterable/ChainedIterator.java:70]:
```suggestion
                    currentIterator = ((Iterable<? extends T>) next).iterator();
```

**Pull request merged**